File: backend/src/main/java/de/grimsi/gameyfin/rest/SetupFilter.java
package de.grimsi.gameyfin.rest;


import de.grimsi.gameyfin.service.SetupService;
import jakarta.servlet.*;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.core.annotation.Order;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Order(1)
@Component
@RequiredArgsConstructor
public class SetupFilter implements Filter {

    private final SetupService setupService;

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws ServletException, IOException {
        HttpServletRequest req = (HttpServletRequest) servletRequest;
        HttpServletResponse res = (HttpServletResponse) servletResponse;

        boolean isSetupUri = req.getRequestURI().contains("/v1/setup");

        if (setupService.isSetupCompleted() && !isSetupUri ||
                !setupService.isSetupCompleted() && isSetupUri) {
            filterChain.doFilter(req, res);
        } else {
            res.setStatus(HttpStatus.FORBIDDEN.value());
        }
    }
}

File: src/main/kotlin/de/grimsi/gameyfin/setup/SetupFilter.kt
package de.grimsi.gameyfin.setup

import jakarta.servlet.*
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import java.io.IOException


@Order(1)
@Component
class SetupFilter(
    private val setupService: SetupService
) : Filter {

    @Throws(ServletException::class, IOException::class)
    override fun doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain) {
        val req = servletRequest as HttpServletRequest
        val res = servletResponse as HttpServletResponse

        val isSetupUri = req.requestURI.startsWith("/setup")
        val isLoginUri = req.requestURI.startsWith("/login")

        // Skip this filter if the urls don't match
        if (!(isSetupUri || isLoginUri)) {
            filterChain.doFilter(req, res)
            return
        }

        val isSetupComplete = setupService.isSetupCompleted()

        if (isSetupUri && isSetupComplete) {
            res.sendRedirect("/login")
        } else if (isLoginUri && !isSetupComplete) {
            res.sendRedirect("/setup")
        }

        // took me longer than I want to admit to realize you always need to call doFilter() at the end
        filterChain.doFilter(req, res)
    }
}